src/backend/distributed/commands/multi_copy.c

```diff
@@ -465,7 +465,7 @@ CopyToExistingShards(CopyStmt *copyStatement, char *completionTag)
 	 * PG_CATCH. Otherwise, it may be undefined in the PG_CATCH (see sigsetjmp
 	 * documentation).
 	 */
-	shardConnectionHash = CreateShardConnectionHash();
+	copyConnectionHash = CreateShardConnectionHash(MessageContext);
```

Comment:
> FWIW, I think we should backpatch that, i.e. fix it individually.

Done, and released this week in 5.2.1.

> We'd have to wait for all of that to complete before resetting the memory context. […] I'm generally loathe to use MessageContext for longer lived stuff.

Changed to `TopTransactionContext`.